site/docs/plugins/parse-mode.md

```diff
@@ -2,27 +2,134 @@
 
 This plugin provides a transformer for setting default `parse_mode`, and a middleware for hydrating `Context` with familiar `reply` variant methods - i.e. `replyWithHTML`, `replyWithMarkdown`, etc.
 
-## Usage
+## Usage (Using format)
 
 <CodeGroup>
   <CodeGroupItem title="TypeScript" active>
 
 ```ts
-import { Bot } from "grammy";
+import { Bot, Composer, Context } from "grammy";
+import { bold, fmt, hydrateReply, italic } from "@grammyjs/parse-mode";
+
+import type { ParseModeFlavor } from "@grammyjs/parse-mode";
+
+const bot = new Bot<ParseModeFlavor<Context>>("");
+
+// Install format reply variant to ctx
+bot.use(hydrateReply);
+
+bot.command("demo", async (ctx) => {
+  await ctx.replyFmt(fmt`${bold("bold!")}
+${bold(italic("bitalic!"))}
+${bold(fmt`bold ${link("blink", "example.com")} bold`)}`);
```

Comment:
Why is this call to `fmt` even necessary? Couldn't we support
```ts
fmt`${bold`${italic`bitalic!`}`}`
```
?

Reply:
(this is off-topic, it's not about the docs)

Reply:
Not quite sure what you mean. The bitalic word doesn't have any enclosing fmt. The next part, bold blink bold, uses fmt to combine string primitives with FormattedString.

Reply:
> The bitalic word doesn't have any enclosing fmt.

Not directly, but it still passes through the outer `fmt` so technically this should be enough, right?

Reply:
In other words, nested `fmt` calls are only necessary because `fmt` lacks support for processing nested formatting entities.

Reply:
> Not directly, but it still passes through the outer `fmt` so technically this should be enough, right?

No, not at all.

> In other words, nested `fmt` calls are only necessary because `fmt` lacks support for processing nested formatting entities.

Not exactly. It's more of a property of tagged template literal, rather than a property of `fmt` and it's implementation.

A template literal can be written as the general form `stringPart1 ${expr1} stringPart2 ${expr2} ... {$exprN} stringPartN+1`. A tagged template literal takes this template literal and call the function with the following signature `tag([stringPart1, ... stringPartN+1], expr1, ... exprN)`.

Then, we have `fmt` defined with the following type signature `fmt: (rawStringParts: TemplateStringsArray | string[], ...stringLikes: Stringable[]) => FormattedString`. That is, whatever expression comes out of `expr_n`, it ought to be of type `Stringable`.

So let us consider the text `bold blink bold` where all the terms are bolded, and the middle term is also a link. Here, we opt to enclose the entire text with `bold("bold blink bold")` first, since everything is bolded. This is still good, as the return type of `bold` is a `Stringable`. Next, we need to linkify the middle element. Of course, we cannot naively do ```bold(`bold ${link("blink", "...")} bold`)``` as this expression needs to resolve to a value before being handed over to the outer `fmt`, and a regular template literal will resolve this to a string - that is, it will try to cast any non-string expressions to string, causing us to lose precious link formatting information for `blink`.

Thus, we do ```bold(fmt`bold ${link("blink", "...")} bold`)``` instead. Here, the inner expression resolves to a `Stringable` that is then fed to the outer `bold` function which returns another `Stringable`. Finally, this entire expression return value is fed to the outermost `fmt` to be concatenated into the complete `Stringable`.

Reply:
That said, if somebody insists on using only one `fmt`, they can. They would just need to call `bold` three times for each individual term in the `bold blink bold` case, with the middle term having an additional `link` call, of course.